Prohibit assigning concurrent maps into Map-types variables and fields

leventov opened the pull request:
Concurrent maps (i. e. `ConcurrentHashMap` and `ConcurrentSkipListMap`) should be assigned into variables of their respective type or `ConcurrentMap`, but not just `Map`.

Why this is important could be seen in `CoordinatorRuleManager`, where it's pretty obvious that code
```java
ConcurrentMap<String, List<Rule>> theRules = rules.get();
if (theRules.get(dataSource) != null) {
  retVal.addAll(theRules.get(dataSource));
}
```
has a race condition, but previously when the type of the variable was `Map` it was not obvious.

This race condition in `CoordinatorRuleManager` is fixed in this PR. Also, improved logic in `DirectDruidClient` and `ResourcePool`.

QiuMM commented on server/src/main/java/org/apache/druid/server/coordinator/ReplicationThrottler.java:
```diff
@@ -161,10 +161,10 @@ public void resetLifetime(String tier)
-      return retVal;
+      ConcurrentMap<SegmentId, String> segments = currentlyProcessingSegments.get(tier);
+      List<String> segmentsAndHosts = new ArrayList<>();
+      segments.forEach((segmentId, serverId) -> segmentsAndHosts.add(segmentId + " ON " + serverId));
```
I wonder why you didn't use `StringUtils.format`.

leventov replied:
Simple string concatenation should be faster because it doesn't involve parsing of the format.

QiuMM replied:
Got it.

leventov commented on server/src/main/java/org/apache/druid/client/DirectDruidClient.java:
```diff
@@ -159,14 +158,15 @@ public int getNumOpenConnections()
 
     Pair<JavaType, JavaType> types = typesMap.get(query.getClass());
     if (types == null) {
```
It could improve concurrency, see https://github.com/apache/incubator-druid/pull/4397#discussion_r122067079. I'll add a comment.

QiuMM replied:
Got it.

QiuMM replied:
Hmm, `computeIfAbsent()` will call `get()` again, so I think we shouldn't use `computeIfAbsent()` here. Just revert this modification.

QiuMM replied:
Hmm, `computeIfAbsent()` may call `get()` again(ConcurrentHashMap doesn't, ConcurrentSkipListMap does), so I think we shouldn't use `computeIfAbsent()` here, using `put()` is ok.

leventov replied:
If two queries of the same new type are run in parallel, there could be a race between them. Maybe it could be tolerated here because computation (body of the lambda) could be run in parallel and re-run for the same type with no harm, but `computeIfAbsent()` is clearer.

leventov commented:
@QiuMM in a newer commit I've added comments and enforced that `ConcurrentHashMap` type is used when needed.

@jihoonson FYI I refactored `Counters` so that now it just holds static utility methods.

gianm commented on .idea/inspectionProfiles/Druid.xml:
```diff
@@ -236,6 +236,38 @@
+        <constraint name="x" within="" contains="" />
+        <constraint name="y" nameOfExprType="java\.util\.concurrent\.ConcurrentMap" expressionTypes="java.util.concurrent.ConcurrentMap" exprTypeWithinHierarchy="true" within="" contains="" />
+      </searchConfiguration>
+      <searchConfiguration name="A ConcurrentHashMap on which compute() is called should be assinged into variables of ConcurrentHashMap type, not ConcurrentMap" text="$x$.compute($y$, $z$)" recursive="true" caseInsensitive="true" type="JAVA">
```
Consider including the rationale in this message: it is not obvious that it's because ConcurrentMap does not guarantee atomicity.

leventov replied:
This field is not assumed to be a message, it's a configuration name, I think I already overuse them. Probably neither desktop IntelliJ nor TeamCity CI are prepared for something multiline in this field.

gianm commented on indexing-service/src/main/java/org/apache/druid/indexing/common/Counters.java:
```diff
@@ -19,54 +19,31 @@
+  public static <K> int incrementAndGetInt(ConcurrentHashMap<K, AtomicInteger> counters, K key)
   {
-    return intCounters.computeIfAbsent(key, k -> new AtomicInteger()).addAndGet(val);
+    // get() before computeIfAbsent() is an optimization to avoid locking in computeIfAbsent() if not needed.
```
Any idea why ConcurrentHashMap does not already employ an optimization like this?

leventov replied:
That's a throughput vs. scalability tradeoff, + lack of information. We are potentially doing two operations instead of one, and avoid locking in some cases instead.

At some sites where `computeIfAbsent()` is actually expected to find the key absent and compute the value most of the time, the `get()` guard just makes things worse.

There is also an area where it's hard for me to say what approach is better, is that when the map is big and `computeIfAbsent()` constitutes significant part of the app's CPU consumption (the bigger the map and the hotter `computeIfAbsent()` call is, the more likely that it's better to not guard `computeIfAbsent()` with `get()`). I think it's never nearly the case on Druid nodes that `computeIfAbsent()` is hot, but I could be wrong.

leventov replied:
From the ConcurrentHashMap's part, it would be useful if `computeIfAbsentMoreScalableButMaybeDoingExtraWork()` existed, where they don't recompute hash bucket twice and just walk the collision chain twice. But it's easy to imagine why such method doesn't exist.

gianm commented:
It looks like some unit tests are failing now with similar messages. Maybe a recent change broke something?

clintropolis commented on server/src/main/java/org/apache/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java:
```diff
@@ -67,7 +67,7 @@
   private ExecutorService listenerExecutor;
 
-  private final Map<NodeType, NodeTypeWatcher> nodeTypeWatchers = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<NodeType, NodeTypeWatcher> nodeTypeWatchers = new ConcurrentHashMap<>();
```
`computeIfAbsent` is called on this one, https://github.com/apache/incubator-druid/blob/101ca9c00f5fd3721abc01add0aa2b81afcf381d/server/src/main/java/org/apache/druid/curator/discovery/CuratorDruidNodeDiscoveryProvider.java#L91

clintropolis commented on server/src/main/java/org/apache/druid/segment/realtime/appenderator/AppenderatorImpl.java:
```diff
@@ -121,7 +121,7 @@
   private final IndexMerger indexMerger;
   private final Cache cache;
-  private final Map<SegmentIdWithShardSpec, Sink> sinks = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<SegmentIdWithShardSpec, Sink> sinks = new ConcurrentHashMap<>();
```
yeah, I guess this one looks like it could just be `ConcurrentMap`
